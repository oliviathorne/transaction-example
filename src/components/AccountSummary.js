// @flow
import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';

const AccountSummaryWrapper = styled.div`
  display: flex;
  align-items: center;
`;

const Paragraph = styled.p`
  margin: 0;
  font-weight: 400;
  font-size: 10pt;
`;

const AccountInfo = styled.div`
  font-weight: 600;
  padding: 0.75rem;
  width: 33%;

  &:first-child {
    text-align: right;
  }
`;

export const PieChart = styled.div`
  display: inline-block;
  min-width: 60px;
  height: 60px;
  border-radius: 50%;
  position: relative;
  background-image: ${(props) => `conic-gradient(
    #ffb524 0,
    #ffb524 ${3.6 * props.percentageLeft}deg,
    #ebebee ${3.6 * props.percentageLeft}deg,
    #ebebee ${3.6 * 100 - props.percentageLeft}deg
  )`};

  &::before {
    display: inline-block;
    background: #ffffff;
    content: '';
    width: 30px;
    height: 30px;
    border-radius: 50%;
    position: absolute;
    top: 15px;
    left: 15px;
  }
`;

type Props = {
  totalAvailable: number,
  totalSent: number,
  startingBalance: number
};

const AccountSummary = (props: Props) => {
  const TOTAL_AVAILABLE = parseFloat(props.totalAvailable)
    .toFixed(2)
    .replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const TOTAL_SENT = parseFloat(props.totalSent)
    .toFixed(2)
    .replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const PERCENTAGE_LEFT = 100 - (props.totalSent / props.startingBalance) * 100;

  return (
    <AccountSummaryWrapper>
      <AccountInfo>
        £{TOTAL_SENT}
        <Paragraph>total sent</Paragraph>
      </AccountInfo>
      <PieChart percentageLeft={PERCENTAGE_LEFT} />
      <AccountInfo>
        £{TOTAL_AVAILABLE}
        <Paragraph>left available</Paragraph>
      </AccountInfo>
    </AccountSummaryWrapper>
  );
};

const mapStateToProps = (state, ownProps) => ({
  totalAvailable: state.form.totalAvailable,
  totalSent: state.form.totalSent,
  startingBalance: state.form.startingBalance
});

export default connect(mapStateToProps)(AccountSummary);
