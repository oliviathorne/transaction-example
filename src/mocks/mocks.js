export const mockState = {
  form: {
    fields: {},
    savedTransactions: [
      {
        name: {
          value: 'Test name'
        },
        email: {
          value: 'email@test.com'
        },
        amount: {
          value: '3000'
        }
      }
    ],
    startingBalance: 18000,
    totalAvailable: 18000,
    totalSent: 0
  }
};
