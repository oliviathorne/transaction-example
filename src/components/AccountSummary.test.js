import { mount } from 'enzyme';
import React from 'react';
import AccountSummary, { PieChart } from './AccountSummary';
import configureStore from 'redux-mock-store';
import { mockState } from './../mocks/mocks';
import { Provider } from 'react-redux';

const mockStore = configureStore();
let store;
let wrapper;

beforeEach(() => {
  store = mockStore(mockState);
  wrapper = mount(
    <Provider store={store}>
      <AccountSummary />
    </Provider>
  );
});

describe('AccountSummary component', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a graph', () => {
    const graph = wrapper.find(PieChart);

    expect(graph.length).toBe(1);
    expect(graph.props().percentageLeft).toEqual(100);
  });
});
