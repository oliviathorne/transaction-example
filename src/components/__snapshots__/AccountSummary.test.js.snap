// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountSummary component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(AccountSummary)>
    <AccountSummary
      dispatch={[Function]}
      startingBalance={18000}
      totalAvailable={18000}
      totalSent={0}
    >
      <styled.div>
        <StyledComponent
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": true,
                "lastClassName": "hKAetY",
                "rules": Array [
                  "
  display: flex;
  align-items: center;
",
                ],
              },
              "displayName": "styled.div",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bdVaJa hKAetY"
          >
            <styled.div>
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htpNat",
                      "isStatic": true,
                      "lastClassName": "exkFRM",
                      "rules": Array [
                        "
  font-weight: 600;
  padding: 0.75rem;
  width: 33%;

  &:first-child {
    text-align: right;
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "render": [Function],
                    "styledComponentId": "sc-htpNat",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-htpNat exkFRM"
                >
                  £
                  0.00
                  <styled.p>
                    <StyledComponent
                      forwardedClass={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": undefined,
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": true,
                            "lastClassName": "hxPXFl",
                            "rules": Array [
                              "
  margin: 0;
  font-weight: 400;
  font-size: 10pt;
",
                            ],
                          },
                          "displayName": "styled.p",
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "p",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <p
                        className="sc-bwzfXH hxPXFl"
                      >
                        total sent
                      </p>
                    </StyledComponent>
                  </styled.p>
                </div>
              </StyledComponent>
            </styled.div>
            <styled.div
              percentageLeft={100}
            >
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bxivhb",
                      "isStatic": false,
                      "lastClassName": "lfirxx",
                      "rules": Array [
                        "
  display: inline-block;
  min-width: 60px;
  height: 60px;
  border-radius: 50%;
  position: relative;
  background-image: ",
                        [Function],
                        ";

  &::before {
    display: inline-block;
    background: #ffffff;
    content: '';
    width: 30px;
    height: 30px;
    border-radius: 50%;
    position: absolute;
    top: 15px;
    left: 15px;
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "render": [Function],
                    "styledComponentId": "sc-bxivhb",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                percentageLeft={100}
              >
                <div
                  className="sc-bxivhb lfirxx"
                />
              </StyledComponent>
            </styled.div>
            <styled.div>
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htpNat",
                      "isStatic": true,
                      "lastClassName": "exkFRM",
                      "rules": Array [
                        "
  font-weight: 600;
  padding: 0.75rem;
  width: 33%;

  &:first-child {
    text-align: right;
  }
",
                      ],
                    },
                    "displayName": "styled.div",
                    "render": [Function],
                    "styledComponentId": "sc-htpNat",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-htpNat exkFRM"
                >
                  £
                  18,000.00
                  <styled.p>
                    <StyledComponent
                      forwardedClass={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": undefined,
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bwzfXH",
                            "isStatic": true,
                            "lastClassName": "hxPXFl",
                            "rules": Array [
                              "
  margin: 0;
  font-weight: 400;
  font-size: 10pt;
",
                            ],
                          },
                          "displayName": "styled.p",
                          "render": [Function],
                          "styledComponentId": "sc-bwzfXH",
                          "target": "p",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                    >
                      <p
                        className="sc-bwzfXH hxPXFl"
                      >
                        left available
                      </p>
                    </StyledComponent>
                  </styled.p>
                </div>
              </StyledComponent>
            </styled.div>
          </div>
        </StyledComponent>
      </styled.div>
    </AccountSummary>
  </Connect(AccountSummary)>
</Provider>
`;
